[8.0][IMP] Modifications in document approval.

luistorresm opened the pull request:
[HU #1050](https://www.vauxoo.com/web#id=1050&view_type=form&model=user.story&menu_id=557&action=726)
CA #1, 2, 3, 4, 6

[IMP][document_page_approval]
    Added field history_state in document page with widget
    kanban_state_selection, to indicate if all the histories related
    with the document are approved.
    Added test to check the new feature.
[ADD][document_page_code]Added module that add field internal code in
    document page model.
    Added in search view the new field code
[ADD][document_page_approval_levels]
    Added module that set approvers in document.page.
    Added model approver.document.page.item with fields user_id and
    approver to allow have levels to approve an history.
    Hide button approve in history when are not approved all items in
    it.
    Added as followers all users setted in the page as committee.
    Send notification to followers when is created a new history.
    Added profiles to users in knowledge
    Added users demo
    Added tests in module

nhomar commented on document_page_approval_levels/README.rst:
```diff
@@ -0,0 +1,65 @@
+=============================
+
+This module:
+  - TO-DO
```
Do not forget finish the doc.

nhomar commented on document_page_approval_levels/models/document_page.py:
```diff
@@ -0,0 +1,41 @@
+    """This class is use to manage Document."""
+    _inherit = 'document.page'
+
+    approver_items_id = fields.One2many(
```
the field has a semantic issue:

s/approver_items_id/approver_item_ids/g

Set a proper help please to explain the meaning functionally.

nhomar commented on document_page_approval_levels/models/document_page_item.py:
```diff
@@ -0,0 +1,60 @@
+            msg = (
+                '<div><b>The page has been approved by %s .</b></div>' %
+                item.user_id.name)
+            self.env['mail.message'].create({
```
Use message_post not .create directly.

Do an rgrep into sale and purchase module to understand the usage.

nhomar commented on document_page_code/models/document_page.py:
```diff
@@ -0,0 +1,18 @@
+    """This class is use to manage Document."""
+    _inherit = 'document.page'
+
+    code = fields.Char('Internal code')
```
Left the string empty, the word Code is enought.

luistorresm replied:
Thanks @nhomar 

But I have the next requirement:
`Agregar a objeto document.page campo "code" type char, el string debe ser "Internal code" y su traducción correspondiente a español "Código interno"`

nhomar commented on document_page_approval_levels/models/document_page.py:
```diff
@@ -0,0 +1,41 @@
+    @api.multi
+    def write(self, vals):
+        """Added that update approvers in histories related to this
+        document when this change."""
```
I prefer a server action instead a overwritting of the write method, can you use that apporach please, ask @JulioSerna for better understanding of that approach.

JulioSerna replied:
You can create one server action with this logic and the server action will be trigger when this module is update

nhomar commented on document_page_approval_levels/models/document_page.py:
```diff
@@ -0,0 +1,41 @@
+                'user_id': item.user_id.id,
+                'history_id': res.id,
+            })
+        return res
```
Tis approach can be done with a automated action alos instead and overwriting of the method which I supose is bringing a recursion problem.

JulioSerna commented on document_page_approval_levels/models/document_page_history.py:
```diff
@@ -0,0 +1,67 @@
+    def update_approver_items(self):
+        """Refresh items by user based in page_id"""
+        for history in self:
+            users_approver = [
```
what about change this by one `filtered`?

JulioSerna commented on document_page_approval_levels/models/document_page_history.py:
```diff
@@ -0,0 +1,67 @@
+            users_approver = [
+                item.user_id.id for item in history.approver_items_id if
+                item.approved]
+            users_page = [
```
What about change to `filtered`?

JulioSerna commented on document_page_approval/document_page_approval.py:
```diff
@@ -245,6 +245,17 @@ def is_approval_required(self, page):
+        return blocked in the document"""
+        for document in self:
+            document.history_state = 'done'
+            no_approveds = [
```
what about change to `filtered`?

luistorresm committed:
[IMP] document_page_approval:
 - Added field history_state in document page with widget
   kanban_state_selection, to indicate if all the histories related
   with the document are approved.
 - Added test to check the new feature.
 - Added boolean to indicate of all histories are approved in tree view,
   and allow filter by these field
 - Added validation that only users approver can change the content in
   histories

[ADD] document_page_code:
 - Added module that add field internal code in document page model.
 - Added in search view the new field code

[ADD] document_page_approval_levels:
 - Added module that set approvers in document.page.
 - Added model approver.document.page.item with fields user_id and
   approver to allow have levels to approve an history.
 - Hide button approve in history when are not approved all items in it.
 - Added as followers all users setted in the page as committee.
 - Send notification to followers when is created a new history.
 - Added profiles to users in knowledge (Approver/Manager)
 - Added users demo
 - Added group in menu configuration for documents
 - Change List Comprehensions by filtered and mapped when was possible.
 - Added server action to avoid overwrite write method
 - Added that when is changed the content in a history:
     - Send a notification to page followers that are a new version to
       approve.
     - Disapprove the items that has already been approved.
 - When the history is approved, the content in the history is updated, but
 - I add a new element in context, to avoid that create a new history by
   the change.
 - Added test to make the user case developed

Commit:
[FIX][document_page_approval_levels] Flake8 error fixed.
